### PR TITLE
[ci, model] test: add bitwise logits-equal tests for transformers v4 models

### PR DESCRIPTION
**Pull request by TimYangst**
### What does this PR do?

> Adds a single-GPU bitwise logits-equality test (`tests/models/test_models_logits_equal.py`) that compares pristine HuggingFace forwards against VeOmni's monkey-patched HF models on transformers v4. Coverage spans causal-LM (qwen2.5, llama3.1, seed_oss, qwen3, qwen3_moe, deepseek_v3), VLM (qwen2_vl, qwen2_5_vl, qwen3_vl, qwen3_vl_moe), and Omni thinker (qwen2_5_omni, qwen3_omni_moe) families. VL and Omni cases run a real multimodal forward — a single dummy 2x2 image — so the visual tower actually executes; the comparison guards against silent numerical drift in both the language and visual stacks.
>
> Two parallel test functions: `test_logits_bitwise_equal` exercises every family via the manual layout adapter in `tests/models/weight_sync_adapters.py`; `test_logits_bitwise_equal_via_runtime_converter` covers the MoE families (qwen3_moe, deepseek_v3) through the `MoEV4StackingConverter` path real users hit when pointing at a vanilla HF checkpoint dir.

### Checklist Before Starting

- Search for relative PRs/issues and link here: N/A
- PR title follows `[{modules}] {type}: {description}` format

### Test

> 26/26 cases (22 forward + 4 runtime-converter) pass on a single A100 in ~38s under the existing `transformers-stable` group (transformers 4.57.3).
>
> CI: `.github/workflows/gpu_unit_tests.yml` already runs `tests/models/` under transformers v4, so the new file is picked up without workflow edits. Module-wide skip gates (`is_transformers_version_greater_or_equal_to("5.0.0")`, `IS_CUDA_AVAILABLE`) make the file a no-op outside that scope, and the autouse process-group fixture mirrors those gates so a transformers v5 environment does not pay an unnecessary NCCL init.

### API and Usage Example

> N/A — test-only change.

### Design & Code Changes

> - **Test scaffolding**. `tests/models/test_models_logits_equal.py` parametrises cases pairing HF toy configs with VeOmni builds. HF runs first (before any VeOmni import that could monkey-patch HF classes), state-dict is copied to VeOmni via `weight_sync_adapters`, then bitwise equality is asserted on the logits.
>
> - **HF class resolution.** No per-case `hf_class` override; `_resolve_hf_class(config)` looks up `config.architectures[0]` against the `transformers` namespace, which works uniformly for plain causal-LMs (`Qwen2ForCausalLM` ...), VLMs (`Qwen2_5_VLForConditionalGeneration` ...), and the Omni wrappers (`Qwen2_5OmniForConditionalGeneration` ...). Construction goes through `cls._from_config(config, torch_dtype=..., attn_implementation=...)` so weight-init RNG sees the requested dtype directly — a `cls(config).to(bf16)` path would init in fp32 first, breaking bitwise equality against VeOmni's bf16-from-the-start build.
>
> - **Forward inputs.** `_make_inputs(case, config, device, dtype)` is the single place that builds input_ids + forward kwargs:
>   - Causal-LM: text-only forward, vocab floor 32000 dodges every 151xxx multimodal placeholder id.
>   - VL / Omni: same base ids, but the first `n_tokens = grid_t * grid_h * grid_w // spatial_merge_size**2` positions are overwritten with `image_token_id` and dummy `pixel_values` + `image_grid_thw = [[1, 2, 2]]` are passed alongside, so the visual tower runs and the patched `masked_scatter` consumes every image embedding exactly once. `pixel_values` shape and dtype are derived from `vision_config` (qwen2/2.5: patch_size=14; qwen3: patch_size=16). `image_mask` / `video_mask` (and `audio_mask` for Omni) are passed explicitly because the VeOmni patches otherwise call `dist.all_gather` on input_ids to recompute them.
>   - Audio (`input_features`) is a follow-up: Omni cases pass zero `audio_mask` to satisfy the patched asserts, but no `input_features` means the audio tower stays dark.
>
> - **Process group fixture.** `_single_rank_process_group` is a module-scoped 1-rank NCCL group required by the qwen2_5_vl patch's unconditional `dist.all_gather`. Before init it binds the accelerator via `get_torch_device().set_device(LOCAL_RANK)` so multi-GPU CI hosts don't have NCCL pick the wrong visible device. Init/destroy is gated on whether the fixture itself created the group, and it mirrors the test bodies' transformers-version + CUDA skip gates so a skip-only environment doesn't pay an init cost.
>
> - **HF class restoration (`tests/models/hf_unpatch.py`).** Extended to cover every `apply_veomni_*_patch()` site for the added VLM / Omni families (qwen2_vl, qwen2_5_vl, qwen3_vl, qwen3_vl_moe, qwen2_5_omni, qwen3_omni_moe). These patches do in-place class swaps on the `transformers.models.*` modules (e.g. `hf_qwen2vl.Qwen2VLForConditionalGeneration = ...`), so without restoring them the eager case for a family would leak its VeOmni-patched classes into the same family's fa2 case, silently turning the bitwise comparison into VeOmni-vs-VeOmni. The helper snapshots pristine attributes at import time and restores class swaps + replaced in-class methods (forward / `_deepstack_process` / `get_rope_index` / `_init_weights` / `_no_split_modules`); methods VeOmni adds (`dummy_forward`, `get_parallel_plan`) are removed via `delattr`.
>
> - **Toy-config fix.** `tests/toy_config/qwen25omni_toy/config.json`: `architectures[0]` was `Qwen2_5OmniModel`, which doesn't exist in transformers 4.57.3; corrected to `Qwen2_5OmniForConditionalGeneration` (the class VeOmni's loader returns for this config anyway). Required by the new config-driven HF class resolution.

### Checklist Before Submitting

- [ ] Read the [Contribute Guide](https://github.com/ByteDance-Seed/VeOmni/blob/main/CONTRIBUTING.md)
- [ ] Applied pre-commit checks
- [ ] Added/updated documentation
- [ ] Added tests to CI workflow (or explained why not feasible)

### CODE REVIEW

File: tests/models/test_models_logits_equal.py [hunk 1]
```diff
@@ -103,32 +261,139 @@ def _apply_determinism():
...
+
+    if not IS_CUDA_AVAILABLE or is_transformers_version_greater_or_equal_to("5.0.0"):
+        yield
+        return
+
+    import torch.distributed as dist
+
+    we_initialised = False
+    if not dist.is_initialized():
+        dist.init_process_group(backend="nccl", rank=0, world_size=1)
```
[the remark this answers is not in the document]
Comment (TimYangst) [reply tests/models/test_models_logits_equal.py]: won't fix — the suggestion as authored breaks the build, and the defensive form it's reaching for is dead code for a 1-rank fixture. If we ever extend this test to multi-GPU, that's the right time to add set_device(LOCAL_RANK).

**Review by Luosuu — APPROVED** [review]
